core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrPatternAlternationOptimizer.java

```diff
@@ -0,0 +1,145 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.rowpattern;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.sql.planner.rowpattern.IrPatternQuantifier.zeroOrOne;
+
+/**
+ * Remove empty pattern from pattern alternation and replace it with quantification of a neighbouring term.
+ */
+public final class IrPatternAlternationOptimizer
+        extends IrRowPatternVisitor<IrRowPattern, Void>
+{
+    public static IrRowPattern optimize(IrRowPattern node)
+    {
+        return new IrPatternAlternationOptimizer().process(node);
+    }
+
+    @Override
+    protected IrRowPattern visitIrRowPattern(IrRowPattern node, Void context)
+    {
+        throw new UnsupportedOperationException("unsupported node type: " + node.getClass().getName());
+    }
+
+    @Override
+    protected IrRowPattern visitIrLabel(IrLabel node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrAnchorPattern(IrAnchorPattern node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrEmptyPattern(IrEmptyPattern node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrExcludedPattern(IrExcludedPattern node, Void context)
+    {
+        IrRowPattern child = process(node.getPattern());
+
+        return new IrExcludedPattern(child);
+    }
+
+    @Override
+    protected IrRowPattern visitIrPatternAlternation(IrPatternAlternation node, Void context)
+    {
+        List<IrRowPattern> children = node.getPatterns().stream()
+                .map(this::process)
+                .collect(toImmutableList());
+
+        List<IrRowPattern> emptyChildren = children.stream()
+                .filter(IrEmptyPattern.class::isInstance)
+                .collect(toImmutableList());
+        if (emptyChildren.isEmpty()) {
+            return new IrPatternAlternation(children);
+        }
+        checkState(emptyChildren.size() == 1 && children.size() > 1, "run IrRowPatternFlattener first to remove redundant empty pattern");
+
+        IrRowPattern emptyChild = getOnlyElement(emptyChildren);
+        int emptyChildIndex = children.indexOf(emptyChild);
+
```

Comment:
It seems like this could all be simplified:

```
int emptyChildIndex = -1;
for (int i = 0; i < children.size(); i++) {
    if (children.get(i) instanceof IrEmptyPattern) {
        emptyChildIndex = i;
        break;
    }
}
```